linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala

```diff
@@ -47,6 +49,7 @@ object Analysis {
    */
   trait ClassInfo {
     def className: ClassName
+    def data: Infos.ClassInfo
```

Comment:
Hmmm... But what is the alternative? We need an info-like thing to determine module dependencies. Are you suggesting we re-calculate this if necessary with another tree traverse?

Reply:
We can potentially extend the `Analysis`, but now I see that there is a bigger problem: `LinkedClass`, which is in `standard`, contains references to `Infos` as well. That goes beyond intention, and directly into stabler-references-less-stable territory, which is not good at all.

From what I see, those Infos are needed in `ModuleAnalyzer.staticDependencies`. IIUC, this computes a `dependencies: Set[ClassName]`. Could we compute these `dependencies` eagerly in the `Infos`/`Analyzer`? Then we could expose that `dependencies` in `Analysis` (where it would make sense) and further in `LinkedClass` (which would therefore not depend on a type less stable than itself).

Reply:
> Could we compute these dependencies eagerly in the Infos/Analyzer?

Yes. The first prototype did exactly that. Downsides:

- It makes the Analyzer even more complicated.
- The Analyzer now contains specifics to modules.
- We calculate it twice but only need it once (if we do not add yet another flag to the `Analyzer`).

I feel like the `dependencies` we calculate in `ModuleAnalyzer` are much more private than the Infos. Notably, they are not independent of other classes and depend on `symbolRequirements`. The Infos on the other hand are a mere summary of the trees of a class.

An alternative to what we have right now, might be to make Info caching more central. As such, both the `Refiner` and the `ModuleAnalyzer` could access Infos (without exposing them in the `LinkedClass`). It is unclear how to properly identify the `LinkedClass` in this case though.

Reply:
If we break down the things that `staticDependencies` computes, there are:

* Things read from `ReachabilityInfo`, among which
  * `instantiatedClasses` and `accessedModules`, which depend on `isNonNative`. Why do we need to know whether it is non-native? Could we get away without this knowledge? If not, we could store them in a separate set that the `ModuleAnalyzer` will filter for non-native classes.
  * Everything else, which do not depend on other classes.
* Things read directly from the `classInfos`, which can also be read from existing fields of `LinkedClass`, AFAICT
* Things depending on the `symbolRequirements`, which are only added to `j.l.Object`, and that the `ModuleInitializer` could patch up when reading the dependencies of `j.l.Object` (after all, that *is* clearly a private implementation detail of the ModuleInitializer).
* The artificial dependency from everything on `j.l.Object`, which the ModuleInitializer can also patch up when reading the dependencies from the `LinkedClass`.

That would remove all the dependencies on other classes and on `symbolRequirements`.

The Infos might be less private in terms of exposed information, but as an API they are very unstable, and clearly an implementation detail of the `Analyzer`.

Reply:
> Why do we need to know whether it is non-native?

Because instantiating a native class does not actually need the SJS module that class is in. But the native module that class is in.

This is why we need the distinction between, for example, `instantiatedClasses` and `methodsCalledStatically.keys`. A static method of a native class is still an SJS thing. However, a new instance is not.

> Could we get away without this knowledge?

No.

So in terms of `ReachabilityInfo` this establishes that we need finer granularity than simply a `Set[ClassName]`. The alternative I see here is that we make a difference between a static and an external dependency in the Analyzer. This is not really a breach of abstraction, since it follows from the IR spec (the distinction, what exactly constitutes a static dependency is, AFAICT an implementation detail of the Emitter, so we probably have a breach of abstraction either way).

The upside of this would be:
- No exposed infos.
- No exposed `isParentDataAccessed`

The downside of this would be:
- More complexity in the Analyzer.
- static and external dependencies are calculated pre optimizer for no reason (if it is enabled).
- static and external dependencies are calculated even without module splitting.

Reply:
@sjrd would you like me to try this? How I expect this to look is that `LinkedClass` would get a `staticDependencies: Set[ClassName]` and a `externalDependencies: Set[String]` field. (but we should probably introduce a more specific type for `externalDependencies`).

Reply:
Yes, I believe that plan sounds good :)

Reply:
I have added a new commit with this. I think it is better.

The only thing I do not like is how we handle module initializers. Maybe we should introduce a `LinkedModuleInitializer`? It is still not super nice, that the `Refiner` is re-classifying entry points into modules.